consensus/state_sync/state_sync.go

```diff
@@ -0,0 +1,104 @@
+	//   - Store Block:    consensusModule.CommitBlock()
+	SetHandleBlock(HandleBlock func(block *typesCons.Block) error)
+
+	// `GetPeerSyncMeta` function contract:
+	// - Retrieve a list of active peers with their metadata (identified and retrieved through P2P's `Churn Management`)
+	SetGetPeerMetadata(GetPeerSyncMeta func() (peers []PeerSyncMeta, err error))
+
+	// `NetworkSend` function contract:
+	// - sends data to an address via P2P network
+	SetNetworkSend(NetworkSend func(data []byte, address cryptoPocket.Address) error)
+
+	// -- Sync modes --
+
+	// In the StateSync protocol, the Node fields valid BlockRequests from its peers to help them CatchUp to be Synced.
+	// This sub-protocol is continuous throughout the lifecycle of StateSync.
+	RunServerMode()
+
+	// In PacemakerMode, the Node is caught up to the latest block and relies on the Consensus Module’s Pacemaker
+	// to maintain a synchronous state with the global SyncState.
+	// - UpdatePeerMetadata from P2P module
+	// - UpdateSyncState
+	// - Defer Sync to Pacemaker
+	// - If `localSyncState.Height < globalNetworkSyncState.Height` -> RunSyncMode() // careful about race-conditions with Pacemaker
+	RunPacemakerMode()
+
+	// Runs sync mode 'service' that continuously runs while `localSyncState.Height < globalNetworkSyncState.Height`
+	// - UpdatePeerMetadata from P2P module
+	// - Retrieve missing blocks from peers
+	// - Process retrieved blocks
+	// - UpdateSyncState
+	// - If `localSyncState.Height == globalNetworkSyncState.Height` -> RunPacemakerMode()
+	RunSyncMode()
+
+	// Returns the `highest priority aka lowest height` missing block heights up to `max` heights
+	GetMissingBlockHeights(state SyncState, max int) (blockHeights []int64, err error)
+
+	// Random selection of eligilbe peers enables a fair distribution of blockRequests over time via law of large numbers
+	// An eligible peer is when `PeerMeta.MinHeight <= blockHeight <= PeerMeta.MaxHeight`
+	GetRandomEligiblePeersForHeight(blockHeight int64) (eligiblePeer PeerSyncMeta, err error)
+
+	// Uses `NetworkSend` to send a `BlockRequestMessage` to a specific peer
+	SendBlockRequest(peerId string) error
+
+	// Uses 'NetworkSend' to send a `BlockResponseMessage` to a specific peer
+	// This function is used in 'ServerMode()'
+	HandleBlockRequest(message BlockRequestMessage) error
+
+	// Uses `HandleBlock` to process retrieved blocks from peers
+	// Must update sync state using `SetMissingBlockHeight`
+	ProcessBlock(block *typesCons.Block) error
+}
+
+type SyncState interface {
+	// latest local height
+	LatestHeight() int64
+	// latest network height (from the aggregation of Peer Sync Meta)
+	LatestNetworkHeight() int64
+	// retrieve peer meta (actively updated through churn management)
+	GetPeers() []PeerSyncMeta
+	// returns ordered array of missing block heights
+	GetMissingBlockHeights() []int64
+	// set missing block height if an error occurred during processing
+	SetMissingBlockHeight(blockHeight int64) error
```

Comment:
What did you have in mind for this?

I feel like the internal state of the module should accurately reflect `LatestHeight` and `GetMissingBlockHeights` depending on how it operates so this shouldn't be necessary.

Reply:
Blocks are set as `!Missing` when received from a peer but before validation to not block the block download loop.

If a received block is invalid - it must `Set` a block height as `Missing` again

Reply:
If we have a package (including a struct) implementing the `StateSyncModule`, I would assume it has an internal state that is modified based on some of the functions exposed (e.g. `HandleBlock`).

This way, the state internals of the implementation knows if a block is missing or not without having an external party modify it. The decision as to whether the module should discard additional `HandleBlock` for a height it has cached, but not processed, is left up to the implementor.

My goal here is to expose as few functions as necessary in the interface so the "contract" is simple, which gives more optionality to the implementor.

Reply:
So basically what you're saying is *this should be internal* - that's fine by me